// Copyright 2021 VMware, all rights reserved
//
// Privacy Wallet's api

syntax = "proto3";
package vmware.concord.privacy.wallet.api.v1; 

import "google/protobuf/duration.proto";
import "google/protobuf/any.proto";

// Privacy Wallet Service Interface
service PrivacyWalletService {
    rpc PrivacyWalletService(PrivacyWalletRequest) returns (PrivacyWalletResponse);
}

message PrivacyAppConfigRequest {
    bool budget = 1;
    uint32 numValidators = 2; // The number of validator shares required to reconstruct a signature
    // These are public keys for replicas to encrypt their secret shares
    // TODO: needs to get integrated with encrypting shares currently shares on UTT configuration are clear
    repeated string validatorPublicKey = 3;
}

message PrivacyAppConfigResponse {
    bytes configuration = 1; // Blob of privacy configuration data
}

message PrivacyWalletConfigRequest {
    string private_key = 1;
    string public_key = 2;
    bytes public_application_config = 3;
    string storage_path = 4;
}

message PrivacyWalletConfigResponse {
    bool succ = 1;
}

message UserRegistrationRequest {
}

message UserRegistrationResponse {
    string pid = 1;
    bytes rcm1 = 2;
    bytes rcm1_sig = 3;
}

message UserRegistrationUpdateRequest {
    bytes rcm_sig = 1;
    repeated uint64 s2 = 2;
}

message UserRegistrationUpdateResponse {
    bool succ = 1;
}

message PrivacyWalletRequest {
    oneof req {
        PrivacyAppConfigRequest privacy_app_config = 1;
        PrivacyWalletConfigRequest privacy_wallet_config_request = 2;
        UserRegistrationRequest user_registration_request = 3;
        UserRegistrationUpdateRequest user_registration_update_request = 4;
    }
    // Required timeout which defines the maximum amount of time the caller is
     // willing to wait for the request to be processed by a quorum of replicas.
    // Returns DEADLINE_EXCEEDED if the request times out.
    // Returns INVALID_ARGUMENT if the timeout is zero.
    google.protobuf.Duration timeout = 5;    
}

message PrivacyWalletResponse {
    optional string err = 1;
    oneof resp {
        PrivacyAppConfigResponse privacy_app_config_response = 2;
        PrivacyWalletConfigResponse privacy_wallet_config_response = 3;
        UserRegistrationResponse user_registration_response = 4;
        UserRegistrationUpdateResponse user_registration_update_response = 5;
    }
}